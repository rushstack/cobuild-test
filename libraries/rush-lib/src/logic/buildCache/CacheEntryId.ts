// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

const OPTIONS_ARGUMENT_NAME: string = 'options';

/**
 * Options for generating the cache id for an operation.
 * @beta
 */
export interface IGenerateCacheEntryIdOptions {
  /**
   * The name of the project
   */
  projectName: string;
  /**
   * The name of the phase
   */
  phaseName: string;
  /**
   * A hash of the input files
   */
  projectStateHash: string;
}

/**
 * Calculates the cache entry id string for an operation.
 * @beta
 */
export type GetCacheEntryIdFunction = (options: IGenerateCacheEntryIdOptions) => string;

const HASH_TOKEN_NAME: string = 'hash';
const PROJECT_NAME_TOKEN_NAME: string = 'projectName';
const PHASE_NAME_TOKEN_NAME: string = 'phaseName';

// This regex matches substrings that look like [token]
const TOKEN_REGEX: RegExp = /\[[^\]]*\]/g;

export class CacheEntryId {
  private constructor() {}

  public static parsePattern(pattern?: string): GetCacheEntryIdFunction {
    if (!pattern) {
      return ({ projectStateHash }) => projectStateHash;
    } else {
      pattern = pattern.trim();

      if (pattern.startsWith('/')) {
        throw new Error('Cache entry name patterns may not start with a slash.');
      }

      const patternWithoutTokens: string = pattern.replace(TOKEN_REGEX, '');
      if (patternWithoutTokens.match(/\]/)) {
        throw new Error(`Unexpected "]" character in cache entry name pattern.`);
      }

      if (patternWithoutTokens.match(/\[/)) {
        throw new Error('Unclosed token in cache entry name pattern.');
      }

      if (!patternWithoutTokens.match(/^[A-z0-9-_\/]*$/)) {
        throw new Error(
          'Cache entry name pattern contains an invalid character. ' +
            'Only alphanumeric characters, slashes, underscores, and hyphens are allowed.'
        );
      }

      let foundHashToken: boolean = false;
      const templateString: string = pattern.trim().replace(TOKEN_REGEX, (token: string) => {
        token = token.substring(1, token.length - 1);
        let tokenName: string;
        let tokenAttribute: string | undefined;
        const tokenSplitIndex: number = token.indexOf(':');
        if (tokenSplitIndex === -1) {
          tokenName = token;
        } else {
          tokenName = token.substr(0, tokenSplitIndex);
          tokenAttribute = token.substr(tokenSplitIndex + 1);
        }

        switch (tokenName) {
          case HASH_TOKEN_NAME: {
            if (tokenAttribute !== undefined) {
              throw new Error(`An attribute isn\'t supported for the "${tokenName}" token.`);
            }

            foundHashToken = true;
            return `\${${OPTIONS_ARGUMENT_NAME}.projectStateHash}`;
          }

          case PROJECT_NAME_TOKEN_NAME: {
            switch (tokenAttribute) {
              case undefined: {
                return `\${${OPTIONS_ARGUMENT_NAME}.projectName}`;
              }

              case 'normalize': {
                return `\${${OPTIONS_ARGUMENT_NAME}.projectName.replace('@','').replace(/\\+/g, '++').replace(/\\/\/g, '+')}`;
              }

              default: {
                throw new Error(`Unexpected attribute "${tokenAttribute}" for the "${tokenName}" token.`);
              }
            }
          }

          case PHASE_NAME_TOKEN_NAME: {
            switch (tokenAttribute) {
              case undefined: {
                throw new Error(
                  'Either the "normalize" or the "trimPrefix" attribute is required ' +
                    `for the "${tokenName}" token.`
                );
              }

              case 'normalize': {
                // Replace colons with underscores.
                return `\${${OPTIONS_ARGUMENT_NAME}.phaseName.replace(/:/g, '_')}`;
              }

              case 'trimPrefix': {
                // Trim the "_phase:" prefix from the phase name.
                return `\${${OPTIONS_ARGUMENT_NAME}.phaseName.replace(/^_phase:/, '')}`;
              }

              default: {
                throw new Error(`Unexpected attribute "${tokenAttribute}" for the "${tokenName}" token.`);
              }
            }
          }

          default: {
            throw new Error(`Unexpected token name "${tokenName}".`);
          }
        }
      });

      if (!foundHashToken) {
        throw new Error(`Cache entry name pattern is missing a [${HASH_TOKEN_NAME}] token.`);
      }

      // eslint-disable-next-line no-new-func
      return new Function(
        OPTIONS_ARGUMENT_NAME,
        `"use strict"\nreturn \`${templateString}\`;`
      ) as GetCacheEntryIdFunction;
    }
  }
}
