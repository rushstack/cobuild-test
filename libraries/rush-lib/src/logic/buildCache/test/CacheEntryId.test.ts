// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { CacheEntryId, GetCacheEntryIdFunction, IGenerateCacheEntryIdOptions } from '../CacheEntryId';

describe(CacheEntryId.name, () => {
  describe('Valid pattern names', () => {
    function validatePatternMatchesSnapshot(
      projectName: string,
      pattern?: string,
      generateCacheEntryIdOptions?: Partial<IGenerateCacheEntryIdOptions>
    ): void {
      const getCacheEntryId: GetCacheEntryIdFunction = CacheEntryId.parsePattern(pattern);
      expect(
        getCacheEntryId({
          projectName,
          projectStateHash: '09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3',
          phaseName: '_phase:compile',
          ...generateCacheEntryIdOptions
        })
      ).toMatchSnapshot(pattern || 'no pattern');
    }

    // prettier-ignore
    it('Handles a cache entry name for a project name without a scope', () => {
      const projectName: string = 'project+name';
      validatePatternMatchesSnapshot(projectName);
      validatePatternMatchesSnapshot(projectName, '[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[phaseName:trimPrefix]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName:normalize]_[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName:normalize]_[phaseName:trimPrefix]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName:normalize]_[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName:normalize]_[phaseName:trimPrefix]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName]_[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName]_[phaseName:trimPrefix]_[hash]');
    });

    // prettier-ignore
    it('Handles a cache entry name for a project name with a scope', () => {
      const projectName: string = '@scope/project+name';
      validatePatternMatchesSnapshot(projectName);
      validatePatternMatchesSnapshot(projectName, '[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[phaseName:trimPrefix]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName:normalize]_[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, '[projectName:normalize]_[phaseName:trimPrefix]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName:normalize]_[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName:normalize]_[phaseName:trimPrefix]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName]_[phaseName:normalize]_[hash]');
      validatePatternMatchesSnapshot(projectName, 'prefix/[projectName]_[phaseName:trimPrefix]_[hash]');
    });
  });

  describe('Invalid pattern names', () => {
    async function validateInvalidPatternErrorMatchesSnapshotAsync(pattern: string): Promise<void> {
      await expect(() => CacheEntryId.parsePattern(pattern)).toThrowErrorMatchingSnapshot();
    }

    it('Throws an exception for an invalid pattern', async () => {
      await validateInvalidPatternErrorMatchesSnapshotAsync('x');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[invalidTag]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('unstartedTag]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[incompleteTag');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[hash:badAttribute]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[hash:badAttribute:attr2]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[projectName:badAttribute]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[projectName:]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[phaseName]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[phaseName:]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[phaseName:badAttribute]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[:attr1]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('[projectName:attr1:attr2]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('/[hash]');
      await validateInvalidPatternErrorMatchesSnapshotAsync('~');
    });
  });
});
