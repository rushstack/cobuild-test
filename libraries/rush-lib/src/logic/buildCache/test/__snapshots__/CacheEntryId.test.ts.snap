// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 1`] = `"Cache entry name pattern is missing a [hash] token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 2`] = `"Unexpected token name \\"invalidTag\\"."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 3`] = `"Unexpected \\"]\\" character in cache entry name pattern."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 4`] = `"Unclosed token in cache entry name pattern."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 5`] = `"An attribute isn't supported for the \\"hash\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 6`] = `"An attribute isn't supported for the \\"hash\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 7`] = `"Unexpected attribute \\"badAttribute\\" for the \\"projectName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 8`] = `"Unexpected attribute \\"\\" for the \\"projectName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 9`] = `"Either the \\"normalize\\" or the \\"trimPrefix\\" attribute is required for the \\"phaseName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 10`] = `"Unexpected attribute \\"\\" for the \\"phaseName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 11`] = `"Unexpected attribute \\"badAttribute\\" for the \\"phaseName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 12`] = `"Unexpected token name \\"\\"."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 13`] = `"Unexpected attribute \\"attr1:attr2\\" for the \\"projectName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 14`] = `"Cache entry name patterns may not start with a slash."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern 15`] = `"Cache entry name pattern contains an invalid character. Only alphanumeric characters, slashes, underscores, and hyphens are allowed."`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: [hash] 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: [phaseName:normalize]_[hash] 1`] = `"_phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: [phaseName:trimPrefix]_[hash] 1`] = `"compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: [projectName:normalize]_[hash] 1`] = `"scope+project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: [projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"scope+project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: [projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"scope+project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: [projectName]_[hash] 1`] = `"@scope/project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: no pattern 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: prefix/[projectName:normalize]_[hash] 1`] = `"prefix/scope+project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: prefix/[projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"prefix/scope+project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: prefix/[projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/scope+project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: prefix/[projectName]_[hash] 1`] = `"prefix/@scope/project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: prefix/[projectName]_[phaseName:normalize]_[hash] 1`] = `"prefix/@scope/project+name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name with a scope: prefix/[projectName]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/@scope/project+name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: [hash] 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: [phaseName:normalize]_[hash] 1`] = `"_phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: [phaseName:trimPrefix]_[hash] 1`] = `"compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: [projectName:normalize]_[hash] 1`] = `"project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: [projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: [projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: [projectName]_[hash] 1`] = `"project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: no pattern 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: prefix/[projectName:normalize]_[hash] 1`] = `"prefix/project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: prefix/[projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"prefix/project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: prefix/[projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: prefix/[projectName]_[hash] 1`] = `"prefix/project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: prefix/[projectName]_[phaseName:normalize]_[hash] 1`] = `"prefix/project+name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names Handles a cache entry name for a project name without a scope: prefix/[projectName]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/project+name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;
